ext/drbbdup/drbbdup.c

```diff
@@ -394,8 +381,15 @@ drbbdup_duplicate_phase(void *drcontext, void *tag, instrlist_t *bb, bool for_tr
     drbbdup_manager_t *manager =
         (drbbdup_manager_t *)hashtable_lookup(&manager_table, pc);
 
-    /* A manager is created if there does not exit one that does not already
-     * book-keeping this bb. */
+    if (!for_trace && manager != NULL && !manager->is_gen) {
+        /* Remove existing invalid book-keeping data. */
+        hashtable_remove(&manager_table, pc);
+        manager = NULL;
+    }
+
+    /* A manager is created if there does not already exit one that "book-keeps"
+     * this basic block.
+     */
```

Comment:
In particular, this part.

Reply:
Yes this is the basic idea.  Does a lock need to be held across the lookup and creation? Thinking of thread-private caches with two threads both creating the same block.  Maybe there's an outer lock I can't see in this diff view.  For that matter I'm wondering whether some of that Dr. Memory code handles thread-private duplicate tags properly...